### PR TITLE
custom VJP by custom JVP and transpose

### PR DESCRIPTION
**Pull request by froystig**
Adds an alternative implementation of `custom_vjp`, behind a flag, that reduces to `custom_jvp` + `custom_transpose`, as sketched out in #9129. Still experimental, not all tests pass yet.

### CODE REVIEW

File: jax/_src/config.py [hunk 1]
```diff
@@ -54,6 +54,14 @@ def int_env(varname: str, default: int) -> int:
   return int(os.getenv(varname, str(default)))
 
 
+UPGRADE_BOOL_HELP = (
+    " This will be enabled by default in future versions of JAX, at which "
+    "point its reverse setting will be considered deprecated. In a version "
```
Comment (mattjj) [inline jax/_src/config.py:59]: Is "reverse setting" clear?

Comment (mattjj) [reply jax/_src/config.py:59]: I wonder if we can just link to a deprecation policy docs page here.

File: jax/_src/config.py [hunk 2]
```diff
@@ -565,6 +578,12 @@ def update_thread_local_jit_state(**kw):
     help=('Select the default PRNG implementation, used when one is not '
           'explicitly provided at seeding time.'))
 
+enable_custom_vjp_by_jvp = config.define_bool_state(
+    name='jax_enable_custom_vjp_by_jvp',
```
Comment (mattjj) [inline jax/_src/config.py:582]:
```suggestion
    name='jax_enable_custom_vjp_by_custom_transpose',
```

**Review by mattjj — APPROVED** [review]

File: jax/_src/config.py (hunk 1 above)
Comment (froystig) [reply jax/_src/config.py]: Good idea. Done.

**Pull request merged**